courses/tasks.py

```diff
@@ -0,0 +1,62 @@
+"""
+Tasks for the courses app
+"""
+import logging
+from mitxpro.celery import app
+from mitxpro.utils import now_in_utc
+from courses.models import CourseRun, CourseRunCertificate
+from courses.utils import ensure_course_run_grade, process_course_run_grade_certificate
+from courseware.api import get_edx_grades_with_users
+
+log = logging.getLogger(__name__)
+
+
+@app.task
+def generate_course_certificates():
+    """
+    Task to generate certificates for courses.
+    """
+    log.info("generate_course_certificates: Started")
+    now = now_in_utc()
+    course_runs = CourseRun.objects.filter(end_date__lt=now).exclude(
+        courseware_id__in=CourseRunCertificate.objects.values_list(
+            "course_run__courseware_id", flat=True
+        )
+    )
+
+    for run in course_runs:
+        edx_grade_user_iter = get_edx_grades_with_users(run)
+
+        for edx_grade, user in edx_grade_user_iter:
+            course_run_grade, created, updated = ensure_course_run_grade(
+                user=user, course_run=run, edx_grade=edx_grade
```

Comment:
@gsidebo `should_update=False` should be the default because the `CourseRunGrade` would exist in case the learner was blacklisted. We do not want to overwrite the blacklist records with the grades received from edX.

Reply:
@pdpinch This is going to need your attention. This is the task to periodically generate certificates for course runs that need them. Unless I'm mistaken, it seems like we will _definitely_ want to fetch and update users' local grade records with the most recent data from edX. @ahmed-arbisoft has pointed out that users' grades might be blacklisted (i.e.: manually set to `0.0`), so we shouldn't update that value. I think this makes it pretty clear that we've overloaded the `grade` value in a problematic way. If we have a new `grade` value from edX, we should update the local record. Our concept of 'blacklisting' should not be preventing this update. I suggest we add a boolean field to `CourseRunGrade` called `blacklisted` or `is_blacklisted`. If that flag is set for some `CourseRunGrade`, we can then prevent any further certificates from being created for that user/run no matter what the grade value is. If we haven't done any actual blacklisting yet, we won't need a data migration. I feel like this should be done as part of this PR

Reply:
Thanks @gsidebo. I agree, my original specification for this was too simplistic. I'm not sure that `is_blacklisted` covers all the use cases though. There are also whitelisted learners who should have a certificate no matter what grade they earned.

Reply:
@pdpinch @ahmed-arbisoft In that case, we should just add a boolean field called `set_by_admin`. If it's `True`, then it should never be updated by edX grade data. Our certificate creation logic already knows what to do based on the `grade` value, whether that value was set by us or by an update from edX grade data